crates/primitives/oracle/src/pragma.rs

```diff
@@ -124,12 +135,45 @@ impl Default for PriceBounds {
     }
 }
 
-fn default_oracle_api_url() -> String {
-    DEFAULT_API_URL.into()
+fn default_oracle_api_url() -> Url {
+    // safe unwrap because its parsed from a const
+    Url::parse(DEFAULT_API_URL).unwrap()
 }
 
 #[derive(Deserialize, Debug)]
 struct PragmaApiResponse {
     price: String,
     decimals: u32,
 }
+
+#[derive(Default)]
+pub struct PragmaOracleBuilder {
+    api_url: Option<Url>,
+    api_key: Option<String>,
+}
+
+impl PragmaOracleBuilder {
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    pub fn with_api_url(mut self, api_url: Url) -> Self {
+        self.api_url = Some(api_url);
+        self
+    }
+
+    pub fn with_api_key(mut self, api_key: String) -> Self {
+        self.api_key = Some(api_key);
+        self
+    }
+
+    pub fn build(self) -> PragmaOracle {
+        PragmaOracle {
+            api_url: self.api_url.unwrap_or_else(default_oracle_api_url),
```

Comment:
Shouldn't it be a failing case when no url is specified?

Reply:
in fact, the builder cannot be reached if url or key is not provided via cli.

removed those useless unwrap and stuff